guides/source/security.md

```diff
@@ -126,6 +126,97 @@ Older versions of Rails use CookieStore, which uses `secret_token` instead of `s
 
 If you have received an application where the secret was exposed (e.g. an application whose source was shared), strongly consider changing the secret.
 
+### Rotating Keys for Encrypted Cookies
+
+The keys, salts, or ciphers used for encrypted cookies can be rotated
+out as needed. This is achieved by using the
+`config.action_dispatch.encrypted_cookie_rotate_out` configuration
+option.
+
+This configuration option can be set to a Hash or an Array of Hashes.
+The Hash must define a `:cipher` value and either a `:secret` or
+`:raw_key` value. If the `:secret` value is defined, a `:salt` value
+must also be defined.
+
+Additionally, when the `:cipher` value is set to AES encryption with CBC
+mode either a `:signed_salt` or `:raw_signed_key` value must also be
+defined in the options Hash. This mode also accepts an optional
+`:digest` value as well.
+
+The difference between the `:secret` and `:raw_key` values is whether
+or not key derivation is performed. Specifically, the `:secret` value
+is treated the same way as the `secret_key_base` and is ran through
+`ActiveSupport::KeyGenerator` with the corresponding `:salt` value.
+Conversely, the `:raw_key` value is *not* run through `KeyGenerator` and
+is instead used directly as the `secret` argument to `MessageEncryptor`.
+
+Suppose we want to rotate out an old `secret_key_base`, the
+`encrypted_cookie_rotate_out` option can be used as follows:
+
+```ruby
+config.action_dispatch.encrypted_cookie_rotate_out = {
+  cipher: "aes-256-gcm",
+  secret: Rails.secrets.old_secret_key_base,
+  salt: "authenticated encrypted cookie"
+}
+```
+
+Multiple rotations are possible through Array of Hashes.
+
+```ruby
+config.action_dispatch.encrypted_cookie_rotate_out = [{
+  # First rotation options...
+}, {
+  # Second rotation options...
+}]
```

Comment:
Agreed it doesn't feel Railsy enough. I like this approach a lot!

Reply:
Cool! There's still some unanswered questions around how to integrate this into Action Dispatch, e.g. the different namespaces `config.rotations.secrets` versus `config.action_dispatch.rotations.cookies`.

Reply:
Yeah I was wondering about the different namespaces. I figured `config.action_dispatch.rotations.cookies` would be the namespace consumed by the cookie middlewares. Can take the various options defined in the configuration and setup the cookies middleware accordingly with the desired rotations. What should the `config.rotations.secrets` namespace be used for?

Additionally, this example uses `:secret_key_base` instead of `:secret`. Should we change the Rotator `:secret` option to be `:secret_key_base`? This is effectively what the option maps to as far as Rails cookies go.

Reply:
Yeah, exactly. I made a comment about how that would look elsewhere.

`config.rotations.secrets` was to rotate `secret_key_base` app-wide. But that likely needs more thought. And needs to look into how `secret_key_base` is used throughout. Come to think of it, it might not even be wrapped in a verifier/encryptor. So… let's shelve that part for now.

I think my usage of `:secret_key_base` above was a confusion on my part. It seems we never pass the secret directly but always generate a secret using the app's `key_generator`.

Reply:
I actually recently published a blog post about the secret key base here: https://medium.com/@mikeycgto/understanding-the-secret-key-base-in-ruby-on-rails-ce2f6f9968a1

As far as I can tell, there are 3 main uses:

1. Encrypted cookies
2. Signed cookies
3. The application message verifier, which has grown in use with ActiveStorage

This PR covers the first 2 cases. I wonder if we should try to cover the last case with these changes?

Reply:
Sweet! Let's leave it to another PR, this touches enough parts already.